### PR TITLE
doc: documentation of flash_area and flash_map

### PR DESCRIPTION
**Pull request by nvlsianpu**
Initial documentation.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

### CODE REVIEW

File: doc/reference/storage/flash_map/flash_mpa.rst [hunk 1]
```diff
@@ -0,0 +1,36 @@
+.. _flash_map_flash_area:
```
Comment (dbkinder) [inline doc/reference/storage/flash_map/flash_mpa.rst]: Looks like the filename is misspelled, and should be ``flash_map.rst``

File: doc/reference/storage/flash_map/flash_mpa.rst [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+Flash map (NVS)
+##########################
+
+Flash map is a way for storing flash partitioning information in one central
+distribution location in flash_area structures array form.
+
+Flash map is generated from DTS based on content of :ref:`flash_partitions`
+nodes.
+I order to use data contain in flash_map the flash_area API is provided.
```
Comment (dbkinder) [inline doc/reference/storage/flash_map/flash_mpa.rst]:
```suggestion
The flash_area API provides a way to access data in the flash map.
```

File: doc/reference/storage/flash_map/flash_mpa.rst [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+flash_area_open() API is the interface for obtaining given flash partitions
```
Comment (dbkinder) [inline doc/reference/storage/flash_map/flash_mpa.rst]:
```suggestion
The flash_area_open() API is the interface for obtaining the flash partitions
```

File: doc/reference/storage/flash_map/flash_mpa.rst [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+flash_area from the flash map.
+
+
+Flash Area API (flash_area)
+###########################
+
+Flash_Area conception combines method for operation on flash chunk
```
Comment (dbkinder) [inline doc/reference/storage/flash_map/flash_mpa.rst]:
```suggestion
The flash_area concept combines methods for operating on a flash chunk
```

File: doc/reference/storage/flash_map/flash_mpa.rst [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+with description of this chunk. Its methods are basically wrappers around
```
Comment (dbkinder) [inline doc/reference/storage/flash_map/flash_mpa.rst]:
```suggestion
together with a description of this chunk. Its methods are basically wrappers around
```

File: doc/reference/storage/flash_map/flash_mpa.rst [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+flash API with addition of input parameters range check and so. Not all flash
```
Comment (dbkinder) [inline doc/reference/storage/flash_map/flash_mpa.rst]:
```suggestion
the flash API, with input parameter range checks. Not all flash
```

File: doc/reference/storage/flash_map/flash_mpa.rst [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+operation are wrapped so API for retrieve flash area driver is included as well.
```
Comment (dbkinder) [inline doc/reference/storage/flash_map/flash_mpa.rst]:
```suggestion
operation are wrapped, so the API for retrieving data using the flash area driver is included as well.
```

File: doc/reference/storage/flash_map/flash_mpa.rst [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+Flash area method are designed to be used along with flash_area structures of
```
Comment (dbkinder) [inline doc/reference/storage/flash_map/flash_mpa.rst]:
```suggestion
The flash area methods are designed to be used along with the flash_area structures of
```

File: doc/reference/storage/flash_map/flash_mpa.rst [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+flash_map and users specific flash_areas with exception of area open API which
```
Comment (dbkinder) [inline doc/reference/storage/flash_map/flash_mpa.rst]:
```suggestion
flash_map and user-specific flash_areas, with the exception of the area_open API
```

File: doc/reference/storage/flash_map/flash_mpa.rst [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+is used to fetch flash_area from flash_map.
```
Comment (dbkinder) [inline doc/reference/storage/flash_map/flash_mpa.rst]:
```suggestion
used to fetch a flash_area from the flash_map.
```

File: include/flash_map.h [hunk 11]
```diff
@@ -33,47 +46,128 @@ extern "C" {
...
  */
 int flash_area_open(u8_t id, const struct flash_area **fa);
 
+/**
+ * @brief Close flash_area
+ *
+ * Reserved for future usage and external projects compatibility reason.
+ * Currently is NOP.
+ *
+ * @param[in] flash_area Flash area to be closed.
```
Comment (dbkinder) [inline include/flash_map.h]: The parameter name is ``fa`` not ``flash_area``, so doxygen is complaining about missing API documentation (lots of times).  See the messages in the shippable log (on the conversatation tab of this PR) for the doc build for this error and others that need fixing in the doxygen comments.

**Review by dbkinder — CHANGES_REQUESTED** [review]

File: doc/reference/storage/flash_map/flash_mpa.rst (hunk 7 above)
Comment (nvlsianpu) [reply doc/reference/storage/flash_map/flash_mpa.rst]: API for retrieve flash area driver - it really mean: to get the appropriate flash driver name which makes possible to operate on the area "by hand" using directly the flash driver API.

Comment (nvlsianpu) [reply doc/reference/storage/flash_map/flash_mpa.rst]: @dbkinder ^^

File: include/flash_map.h [hunk 12]
```diff
@@ -33,47 +46,128 @@ extern "C" {
...
+ * @brief Retrieve partitions flash area form the flash_mpa.
+ *
+ * Function Retrieves flash_area form flash_map for given partition.
+ *
+ * @param[in]  id   ID of the flash partition.
+ * @param[out] fap  Pointer which has to refernce flash_area. If
+ * @p ID is unknown, it will be NULL on output.
  */
-int flash_area_open(u8_t id, const struct flash_area **fa);
+int flash_area_open(u8_t id, const struct flash_area **fap);
```
Comment (carlescufi) [inline include/flash_map.h]: Please don't use `p` for pointer, just stick to `fa`. We don't use Hungarian notation

File: doc/reference/storage/flash_map/flash_map.rst [hunk 13]
```diff
@@ -0,0 +1,36 @@
+.. _flash_map_flash_area:
+
+Flash map (NVS)
+##########################
+
+Flash map is a way for storing flash partitioning information in one central
+distribution location in flash_area structures array form.
```
Comment (carlescufi) [inline doc/reference/storage/flash_map/flash_map.rst]: remove "distribution"

File: doc/reference/storage/flash_map/flash_map.rst [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+flash_area from the flash map.
+
+
+Flash Area API (flash_area)
+###########################
+
+The flash_area concept combines methods for operating on a flash chunk
+together with a description of this chunk. Its methods are basically wrappers around
+the flash API, with input parameter range checks. Not all flash
+operation are wrapped so API for retrieve flash area driver is included as well.
```
Comment (carlescufi) [inline doc/reference/storage/flash_map/flash_map.rst]: "so an API call to retrieve the flash area driver is included as well"

File: include/flash_map.h [hunk 15]
```diff
@@ -5,9 +5,22 @@
...
+ * @file
+ * @brief Public API for flash map
+ */
+
 #ifndef ZEPHYR_INCLUDE_FLASH_MAP_H_
 #define ZEPHYR_INCLUDE_FLASH_MAP_H_
 
+/**
+ * @brief Abstraction over flash area ind its driver which helps to operate on
+ * flash regions gracefully.
```
Comment (carlescufi) [inline include/flash_map.h]: s/gracefully/centrally

File: include/flash_map.h [hunk 16]
```diff
@@ -82,7 +176,7 @@ int flash_area_get_sectors(int fa_id, u32_t *count,
  * Check whether given flash area has supporting flash driver
  * in the system.
  *
- * @param fa Flash area.
+ * @param fa  Flash area.
```
Comment (carlescufi) [inline include/flash_map.h]: why the extra space?

File: include/flash_map.h [hunk 17]
```diff
@@ -91,7 +185,7 @@ int flash_area_has_driver(const struct flash_area *fa);
 /**
  * Get driver for given flash area.
  *
- * @param fa Flash area.
+ * @param fa  Flash area.
```
Comment (carlescufi) [inline include/flash_map.h]: why the extra space?

**Review by carlescufi — CHANGES_REQUESTED** [review]

File: include/flash_map.h (hunk 15 above)
Comment (nvlsianpu) [reply include/flash_map.h]: changed to: `easily and effectively.`

**Comment by nvlsianpu** [comment]
@carlescufi @dbkinder revisit

**Review by carlescufi — APPROVED** [review]

File: doc/reference/storage/flash_map/flash_map.rst [hunk 18]
```diff
@@ -0,0 +1,37 @@
+.. _flash_map_flash_area:
+
+Flash map (NVS)
```
Comment (nvlsianpu) [inline doc/reference/storage/flash_map/flash_map.rst]: ~~NVS~~

**Review by dbkinder — APPROVED** [review]
Doc changes look good, thanks!

**Pull request merged**